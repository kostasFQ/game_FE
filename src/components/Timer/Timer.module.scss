.timersContainer {
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: flex-end;
  box-sizing: border-box;
  height: 10em;
  width: 100vw;
  font-weight: bolder;
  font-size: 20px;

  &__digits {
    position: relative;
    color: red;
  }

  &__counter {
    
    &_visible { display: block }

    &_hide { display: none }
  }
}