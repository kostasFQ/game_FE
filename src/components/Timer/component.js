import React, { Component } from 'react';
import cn from 'classnames';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import styles from './Timer.module.scss';

class Timer extends Component {
  state = {
    seconds: undefined,
    id: undefined
  }

  timerRun() {
    this.interval = setInterval(() => {
      this.setState(state => {
        return { seconds: state.seconds - 1 };
      });
    }, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  componentDidUpdate(prevProps) {
    const { seconds, id } = this.state;
    const { game: { gameOver, gameStarted }, toggleGameOver, setGameTimer, location: { pathname } } = this.props;
    const { game: { gameStarted: prevGameStarted } } = prevProps;
    let [, , sec] = pathname.split('/');

    if (gameStarted && !gameOver && gameStarted !== prevGameStarted) {
      this.timerRun();
    }

    if (gameStarted && seconds === 0 && !gameOver) {
      clearInterval(this.interval);
      toggleGameOver(true);

      this.setState(() => ({ seconds: id }));
      setGameTimer(id);
    }

    if (pathname && pathname.includes('game') && id !== sec) {
      this.setState(() => ({ seconds: sec, id: sec }));
    }
  }

  render() {
    const { seconds } = this.state;
    const { game: { gameOver, gameStarted } } = this.props;

    return (
      <div className={styles.timersContainer}>
        <div className={cn(gameStarted && !gameOver ? styles.timersContainer__counter_visible : styles.timersContainer__counter_hide)}>
          <span className={styles.timersContainer__digits}>{seconds}</span> seconds remain
        </div>
      </div>
    );
  }
}

Timer.defaultProps = {
  gameStarted: false,
  gameOver: false,
  pathname: '',
};

Timer.propTypes = {
  gameStarted: PropTypes.bool,
  gameOver: PropTypes.bool,
  pathname: PropTypes.string,
  setGameTimer: PropTypes.func.isRequired,
  toggleGameOver: PropTypes.func.isRequired,
}

export default withRouter(Timer);
