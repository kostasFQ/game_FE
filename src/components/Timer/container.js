import { connect } from 'react-redux';
import { toggleGameOver, setGameTimer } from 'store/game/actions';
import Timer from './component';

const mapStateToProps = state => ({
  game: state.game,
});

const mapDispatchToProps = dispatch => ({
  toggleGameOver: bool => dispatch(toggleGameOver(bool)),
  setGameTimer: seconds => dispatch(setGameTimer(seconds)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Timer);