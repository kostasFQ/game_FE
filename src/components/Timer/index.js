import Timer from './TimerContainer';
export default Timer;