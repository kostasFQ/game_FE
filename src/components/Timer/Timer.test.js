import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { StaticRouter } from 'react-router-dom';
import _store from '../../../mocks/store';
import Timer from './container';
const mockStore = configureStore([thunk]);

test('Timer should be render', async () => {
  const initialState = { game: _store };
  const store = mockStore(initialState);
  
  const instance = await renderer.create(
    <Provider store={store}>
      <StaticRouter location="/game/1">
        <Timer />
      </StaticRouter>
    </Provider>,
  );

  let component = instance.toJSON();
  expect(component).toMatchSnapshot();
})

test('Timer update seconds', async () => {
  const initialState = { game: _store };
  let store = mockStore(initialState);
  
  const instance = await renderer.create(
    <Provider store={store}>
      <StaticRouter location="/game/1">
        <Timer />
      </StaticRouter>
    </Provider>,
  );

  const updatedState = {
    game: {
      ..._store,
      gameStarted: true,
      gameOver: false
    }
  };
  store = mockStore(updatedState);

  instance.update(
    <Provider store={store}>
      <StaticRouter location="/game/1">
        <Timer />
      </StaticRouter>
    </Provider>,
  )

  let component = instance.toJSON();
  expect(component).toMatchSnapshot();
})

test('Timer update state', async () => {
    const initialState = {
    game: {
      ..._store,
      gameStarted: true,
      gameOver: false,
    }
  };
  let store = mockStore(initialState);

  const instance = await renderer.create(
    <Provider store={store}>
      <StaticRouter location="/game/0">
        <Timer />
      </StaticRouter>
    </Provider>,
  );

  const time = () => new Promise( (res, rej) => {
    setTimeout(() => {
      res("result");
    }, 1000)
  } )
  await time();

  let component = instance.toJSON();
  expect(component).toMatchSnapshot();
})

test('Timer unmount', async () => {
  const initialState = { game: _store };
  let store = mockStore(initialState);
  
  const instance = await renderer.create(
    <Provider store={store}>
      <StaticRouter location="/game/1">
        <Timer />
      </StaticRouter>
    </Provider>,
  );
  instance.unmount();

  let component = instance.toJSON();
  expect(component).toMatchSnapshot();
})