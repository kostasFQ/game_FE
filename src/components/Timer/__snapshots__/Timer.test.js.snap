// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer  1`] = `
<div
  className="timersContainer"
>
  <div
    className="timersContainer__counter_visible"
  >
    <span
      className="timersContainer__digits"
    />
     seconds remain
  </div>
</div>
`;

exports[`Timer should be render 1`] = `
<div
  className="timersContainer"
>
  <div
    className="timersContainer__counter_hide"
  >
    <span
      className="timersContainer__digits"
    />
     seconds remain
  </div>
</div>
`;

exports[`Timer unmount 1`] = `null`;

exports[`Timer update seconds 1`] = `
<div
  className="timersContainer"
>
  <div
    className="timersContainer__counter_visible"
  >
    <span
      className="timersContainer__digits"
    >
      1
    </span>
     seconds remain
  </div>
</div>
`;
